### PR TITLE
DeepSparse server to accept UploadFiles 

### PR DESCRIPTION
**Pull request by dbogunowicz**
Tested on local swagger UI server.

When passing a single image of a goldfish I get the response:

```
{
  "labels": [
    392
  ],
  "scores": [
    12.067000389099121
  ]
}
```

which corresponds to ImageNet's `rock beauty, Holocanthus tricolor'`, a fish very similar to goldfish, so I guess this validates the result (haha).

Cannot process more than one image atm, see the bug described below.

### CODE REVIEW

File: src/deepsparse/image_classification/schemas.py [hunk 1]
```diff
@@ -33,6 +33,10 @@ class ImageClassificationInput(BaseModel):
     class Config:
         arbitrary_types_allowed = True
 
+    @classmethod
+    def from_files(cls, files):
+        return cls([file.filename for file in filenames])
```
Comment (bfineran) [inline src/deepsparse/image_classification/schemas.py:38]:
```suggestion
        return cls(images=[file.filename for file in filenames])
```

File: src/deepsparse/server/main.py [hunk 2]
```diff
@@ -158,17 +158,31 @@ def _add_pipeline_route(
...
-            pipeline,
-            request,
+    if path == "/invocations":
+        @app.post(
+            path,
+            response_model=pipeline.output_schema,
+            tags=["prediction"],
+        )
+        async def _predict_func(files: List[UploadFile]):
+            request = pipeline.output_schema.from_files(files)
```
Comment (bfineran) [inline src/deepsparse/server/main.py:168]:
```suggestion
            request = pipeline.input_schema.from_files(files)
```

**Review by bfineran — CHANGES_REQUESTED** [review]
let's change the logic here so that 
1) if `pipeline.input_schema` supports `from_files`:
        a) in SageMaker mode we only have 1 route - `/invocations` using files as input
        b) in regular mode we have 2 routes - `"{path}"` and `"{path}/from_files"` that take the `input_schema` and files respectively as input

2) if if `pipeline.input_schema` does not support `from_files`, existing functionality is the same - we only have 1 route per model that takes `input_schema`

to make this easier, I'd suggest adding in a helper function to build the route based on path, pipeline, and target input schema that we can call multiple times if needed

File: src/deepsparse/image_classification/schemas.py [hunk 3]
```diff
@@ -33,6 +33,10 @@ class ImageClassificationInput(BaseModel):
     class Config:
         arbitrary_types_allowed = True
 
+    @classmethod
+    def from_files(cls, files):
```
Comment (bfineran) [inline src/deepsparse/image_classification/schemas.py:37]: let's also change the type of `files` here to standardize to `List[str]`, that way we don't need to depend on `UploadFile` in non `server/` submodule code

**Comment by dbogunowicz** [comment]
Also, it seems that the logic for fetching the actual image is needed (as described here: https://stackoverflow.com/questions/63048825/how-to-upload-file-using-fastapi). on it.

File: src/deepsparse/yolo/schemas.py [hunk 4]
```diff
@@ -41,6 +45,11 @@ class YOLOInput(BaseModel):
         description="List of Images to process"
     )
 
+    @classmethod
+    def from_bytes(cls, bytes: List[bytes]):
+        images = [np.array(Image.open(io.BytesIO(byte_img))) for byte_img in bytes]
```
Comment (dbogunowicz) [inline src/deepsparse/yolo/schemas.py:50]: I think that I would pass bytes to the pipeline itself and perform the processing of bytes into numpy array in there. Also, using PIL here (which is not the deepsparse dependency) for quick prototyping, would change it to reduce dependencies.

**Comment by dbogunowicz** [comment]
As discussed with @rahul-tuli, I am experiencing the following bug, which is probably not coming from my changes.
To reproduce:

I set batch size explicitely to two (as @rahul-tuli pointed out is required)
image

Batch size remains 1 in the generator of src/deepsparse/pipeline::Pipeline(ABC) 
image


In swagger UI I pass two images, which finally give me following output from `src/deepsparse/image_classification/pipelines:: process_inputs`:
image

...error message:

```
Traceback (most recent call last):
  File "/home/damian/deepsparse_venv/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py", line 366, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/home/damian/deepsparse_venv/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
    return await self.app(scope, receive, send)
  File "/home/damian/deepsparse_venv/lib/python3.8/site-packages/fastapi/applications.py", line 269, in __call__
    await super().__call__(scope, receive, send)
  File "/home/damian/deepsparse_venv/lib/python3.8/site-packages/starlette/applications.py", line 124, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/damian/deepsparse_venv/lib/python3.8/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/home/damian/deepsparse_venv/lib/python3.8/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/home/damian/deepsparse_venv/lib/python3.8/site-packages/starlette/exceptions.py", line 93, in __call__
    raise exc
  File "/home/damian/deepsparse_venv/lib/python3.8/site-packages/starlette/exceptions.py", line 82, in __call__
    await self.app(scope, receive, sender)
  File "/home/damian/deepsparse_venv/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
    raise e
  File "/home/damian/deepsparse_venv/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "/home/damian/deepsparse_venv/lib/python3.8/site-packages/starlette/routing.py", line 670, in __call__
    await route.handle(scope, receive, send)
  File "/home/damian/deepsparse_venv/lib/python3.8/site-packages/starlette/routing.py", line 266, in handle
    await self.app(scope, receive, send)
  File "/home/damian/deepsparse_venv/lib/python3.8/site-packages/starlette/routing.py", line 65, in app
    response = await func(request)
  File "/home/damian/deepsparse_venv/lib/python3.8/site-packages/fastapi/routing.py", line 227, in app
    raw_response = await run_endpoint_function(
  File "/home/damian/deepsparse_venv/lib/python3.8/site-packages/fastapi/routing.py", line 160, in run_endpoint_function
    return await dependant.call(**values)
  File "/home/damian/deepsparse_copy/src/deepsparse/server/main.py", line 154, in _predict_func
    results = await execute_async(
  File "/home/damian/deepsparse_copy/src/deepsparse/server/asynchronous.py", line 74, in execute_async
    result = await _LOOP.run_in_executor(_THREADPOOL, pipeline)
  File "/usr/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/damian/deepsparse_copy/src/deepsparse/pipeline.py", line 156, in __call__
    engine_outputs: List[numpy.ndarray] = self.engine(engine_inputs)
  File "/home/damian/deepsparse_copy/src/deepsparse/engine.py", line 215, in __call__
    return self.run(inp, val_inp)
  File "/home/damian/deepsparse_copy/src/deepsparse/engine.py", line 319, in run
    self._validate_inputs(inp)
  File "/home/damian/deepsparse_copy/src/deepsparse/engine.py", line 504, in _validate_inputs
    raise ValueError(
ValueError: array batch size of 2 must match the batch size the model was instantiated with 1
```

**Comment by bfineran** [comment]
@dbogunowicz pushed changes discussed to standardize to files for this iteration.  was unable to reproduce your error and was able to create a batch size 2 model multiple ways using the server api. going to chalk this up to user error and say we proceed, but let's keep an eye out for this 

```
deepsparse.engine.Engine:
        onnx_file_path: /home/benjamin/.cache/sparsezoo/4ce8e2fe-29a1-4027-bead-408d28702643/model.onnx
        batch_size: 2
        num_cores: 16
        num_streams: 0
        scheduler: Scheduler.multi_stream
        cpu_avx_type: avx512
        cpu_vnni: False
```

**Review by markurtz — APPROVED** [review]

**Review by rahul-tuli — APPROVED** [review]

**Pull request merged**